File: examples/Haskell/tripping.hs
{-
    Round-trip demonstrator.
    Runs as a single process for easier testing.
    The client task signals to main when it's ready.
-}
import System.ZMQ4

import Control.Concurrent
import Control.Monad (forM_, forever)
import Data.ByteString.Char8 (pack, unpack)
import Data.Time.Clock (getCurrentTime, diffUTCTime)


client_task :: Context -> Socket Pair -> IO ()
client_task ctx pipe =
    withSocket ctx Dealer $ \client -> do
        connect client "tcp://localhost:5555"
        putStrLn "Setting up test..."
        threadDelay $ 100 * 1000

        putStrLn "Synchronous round-trip test..."
        start <- getCurrentTime
        forM_ [0..10000] $ \_ -> do
            send client [] (pack "hello")
            receive client
        end <- getCurrentTime
        let elapsed_ms = (diffUTCTime end start) * 1000
        putStrLn $ " " ++ (show $ (1000 * 10000) `quot` (round elapsed_ms)) ++ " calls/second"

        putStrLn "Asynchronous round-trip test..."
        start <- getCurrentTime
        forM_ [0..10000] $ \_ -> do
            send client [] (pack "hello")
        forM_ [0..10000] $ \_ -> do
            receive client
        end <- getCurrentTime
        let elapsed_ms = (diffUTCTime end start) * 1000
        putStrLn $ " " ++ (show $ (1000 * 10000) `quot` (round elapsed_ms)) ++ " calls/second"

        send pipe [] (pack "done")

worker_task :: IO ()
worker_task = 
    withContext $ \ctx ->
        withSocket ctx Dealer $ \worker -> do
            connect worker "tcp://localhost:5556"

            forever $ do
                msg <- receive worker
                send worker [] msg

broker_task :: IO ()
broker_task =
    withContext $ \ctx ->
        withSocket ctx Dealer $ \frontend ->
        withSocket ctx Dealer $ \backend -> do
            bind frontend "tcp://*:5555"
            bind backend "tcp://*:5556"

            proxy frontend backend Nothing

main :: IO ()
main =
    withContext $ \ctx ->
        withSocket ctx Pair $ \pipeServer ->
        withSocket ctx Pair $ \pipeClient -> do
            bind pipeServer "inproc://my-pipe"
            connect pipeClient "inproc://my-pipe"

            forkIO (client_task ctx pipeClient)
            forkIO worker_task
            forkIO broker_task

            signal <- receive pipeServer
            putStrLn $ unpack signal

File: examples/C#/tripping.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Runtime.Remoting.Messaging;
using System.Text;
using System.Threading;
using ZeroMQ;

namespace Examples
{
    static partial class Program
    {
        //  Round-trip demonstrator
        //  While this example runs in a single process, that is just to make
        //  it easier to start and stop the example. The client task signals to
        //  main when it's ready.
        public static void Tripping(string[] args)
        {
            CancellationTokenSource cancellor = new CancellationTokenSource();
            Console.CancelKeyPress += (s, ea) =>
            {
                ea.Cancel = true;
                cancellor.Cancel();
            };

            using (ZContext ctx = new ZContext())
            {
                using (var client = new ZActor(ctx, Tripping_ClientTask))
                {
                    (new Thread(() => Tripping_WorkerTask(ctx))).Start();
                    (new Thread(() => Tripping_BrokerTask(ctx))).Start();
                    client.Start();
                    using (var signal = client.Frontend.ReceiveFrame())
                        if (Verbose)
                            signal.ToString().DumpString();
                }
            }
        }



        static void Tripping_ClientTask(ZContext ctx, ZSocket pipe, CancellationTokenSource cancellor, object[] args)
        {
            using (ZSocket client = new ZSocket(ctx, ZSocketType.DEALER))
            {
                client.Connect("tcp://127.0.0.1:5555");
                "Setting up test...".DumpString();
                Thread.Sleep(100);

                int requests;
                "Synchronous round-trip test...".DumpString();
                var start = DateTime.Now;
                Stopwatch sw = Stopwatch.StartNew();
                for (requests = 0; requests < 10000; requests++)
                {
                    using (var outgoing = new ZFrame("hello"))
                    {
                        client.Send(outgoing);
                        using (var reply = client.ReceiveFrame())
                        {
                            if (Verbose)
                                reply.ToString().DumpString();
                        }
                    }
                }
                sw.Stop();
                " {0} calls - {1} ms => {2} calls / second".DumpString(requests, sw.ElapsedMilliseconds, requests * 1000 / sw.ElapsedMilliseconds);

                "Asynchronous round-trip test...".DumpString();
                sw.Restart();
                for (requests = 0; requests < 100000; requests++)
                    using (var outgoing = new ZFrame("hello"))
                        client.SendFrame(outgoing);

                for (requests = 0; requests < 100000; requests++)
                    using (var reply = client.ReceiveFrame())
                        if (Verbose)
                            reply.ToString().DumpString();
                sw.Stop();
                " {0} calls - {1} ms => {2} calls / second".DumpString(requests, sw.ElapsedMilliseconds, requests * 1000 / sw.ElapsedMilliseconds);
                using (var outgoing = new ZFrame("done"))
                    pipe.SendFrame(outgoing);
            }
        }

        //  .split worker task
        //  Here is the worker task. All it does is receive a message, and
        //  bounce it back the way it came:
        static void Tripping_WorkerTask(ZContext ctx)
        {
            using (var worker = new ZSocket(ctx, ZSocketType.DEALER))
            {
                worker.Connect("tcp://127.0.0.1:5556");

                while (true)
                {
                    ZError error;
                    ZMessage msg = worker.ReceiveMessage(out error);
                    if (error == null && worker.Send(msg, out error))
                        continue;
                    // errorhandling, context terminated or sth else
                    if (error.Equals(ZError.ETERM))
                        return; // Interrupted
                    throw new ZException(error);
                }
            }
        }

        //  .split broker task
        //  Here is the broker task. It uses the {{zmq_proxy}} function to switch
        //  messages between frontend and backend:
        static void Tripping_BrokerTask(ZContext ctx)
        {
            using (var frontend = new ZSocket(ctx, ZSocketType.DEALER))
            using (var backend = new ZSocket(ctx, ZSocketType.DEALER))
            {
                frontend.Bind("tcp://*:5555");
                backend.Bind("tcp://*:5556");

                ZError error;
                if (!ZContext.Proxy(frontend, backend, out error))
                {
                    if (Equals(error, ZError.ETERM))
                        return; // Interrupted
                    throw new ZException(error);
                }
            }
        }
    }
}
